src/stack.rs

```diff
@@ -1,14 +1,98 @@
-// Alex Ozdemir <aozdemir@hmc.edu> // <- Your name should replace this line!
+// Adam Dunlap <adunlap@hmc.edU>
 // Starter code for HMC's MemorySafe, week 1
 //
 // The implementation of SinglyLinkedList
 
 use Stack;
+use std::mem;
 
 pub struct SinglyLinkedList<T> {
-    data: T // Probably not what you want
+    sz: usize,
+    fst: Node<T>,
+}
+
+enum Node<T> {
+    Nil,
+    Cons(T, Box<Node<T>>),
 }
 
 impl<T: Eq> Stack<T> for SinglyLinkedList<T> {
-    // ????
+    fn new() -> Self {
+        SinglyLinkedList{sz: 0, fst: Node::Nil}
+    }
+    fn push_front(&mut self, item: T) {
+        // Take ownership of the rest of the list by replacing it with Nil
+        let oldfst = mem::replace(&mut self.fst, Node::Nil);
+
+        self.fst = Node::Cons(item, Box::new(oldfst));
+        self.sz += 1;
+    }
+    fn pop_front(&mut self) -> Option<T> {
+        // Take ownership of the whole list
+        let oldfst = mem::replace(&mut self.fst, Node::Nil);
+        match oldfst {
+            Node::Nil => None,
+            Node::Cons(ret, newfst) => {
+                self.sz -= 1;
+                // Put the rest of the list back into self
+                mem::replace(&mut self.fst, *newfst);
+                Some(ret)
+            }
+        }
+    }
+    fn peek_front(&self) -> Option<&T> {
+        match self.fst {
+            Node::Nil => None,
+            Node::Cons(ref ret, _) => Some(ret)
+        }
+    }
+    fn len(&self) -> usize {
+        self.sz
+    }
+
+    fn remove_first(&mut self, to_remove: &T) -> Option<T> {
+        // Remove an element by pushing and popping all but the first instance
+        // of the element we want to remove which reverses it. Then, reverse it
+        // again.
```

Comment:
This is perhaps less satisfying than a traversal of some kind, but is a pretty neat way to do it none the less.